### ISSUE TITLE
Move sensitive config data to separate file

### ISSUE THREAD

**Issue by claudiuslollarius** [acidanthera/bugtracker#717]
Opencore's `config.plist` file contains data that is system specific and should not be reused, such as:

- MLB
- SystemSerialNumber
- SystemUUID

This makes putting one's working config to a Git repo for sharing and versioning more difficult than it needs to be, since the config file needs to be sanitized before committing.

Would it be possible to have OpenCore read this information from a separate plist file? Doing so would allow for just adding the config file with sensitive data to `.gitignore`

**Comment by vit9696** [comment]
Currently it is not an option, but let's say it is, how should it work with vaulting?

**Comment by E-werd** [comment]
Considering that **config.plist** is already hashed for the vault, I don't think vaulting would need to work any differently.  You could create a second file named **platform.plist** to hold the current *PlatformInfo* section from **config.plist**.  Create the vault as you normally would, create_vault.sh already creates hashes for all the files in the OC folder almost indiscriminately.

It would make configuration more portable in a logistical sense.  It could potentially make support more secure by preventing people from exposing their sensitive data to the community.  It might increase setup complexity a little bit, but that's generally information that doesn't need to be modified once it's set.

All that said, I'm unsure how it could affect other parts of the project.

**Comment by vit9696** [comment]
Signing **platform provisioning data** (let's say it is in that platform.plist file with SN/MLB/ROM/UUID) with the same key does make some sense, but I believe this approach does not solve the problem I talk about. It does not fully apply here, but let me try to explain the details in industrial terms to make the idea more familiar to a wider audience.

Imagine you (**software user**) want to use OpenCore bootloader developed by Acidanthera (**software engineering team**) and configured by X for your system (**software integrator**). So the process looks as follows:

1. _Software engineering team_ prepares releases of OpenCore, kernel extensions, tools, etc.
2. _Software integrator_ develops a configuration that incorporates these releases.
3. **Party A** vaults configuration (EFI contents are signed with **vaulting key**, which is embedded into OpenCore.efi).
4. **Party B** signs configuration (OpenCore.efi is signed with **UEFI SecureBoot key**).
5. _Software user_ installs this configuration.
6. Firmware boots signed OpenCore.efi.
7. Vaulted OpenCore.efi reads signed config.plist.

Originally _software user_ has his firmware provisioned with the public part of **UEFI SecureBoot key**, this process is out of the scope of this explanation. _Software integrator_ at step 2 should also ensure that releases are made by _software engineering team_ by verifying some signature or using a trusted repository, but let's also ignore the exact process of this.

Now we need to understand who _Party A_ and _Party B_ can be.

- We (_software engineering team_) will not sign anything.
- Vaulting key is embedded into OpenCore.efi, so vaulting (_Party A_) must happen before signing OpenCore.efi (_Party B_).
- I do not think it makes good sense to store private keys (keys allowing you to create new vault configurations or sign new bootloaders) on the same machine when it is possible to avoid. This somewhat reduces the security by making these keys easier to compromise. Therefore _software user_ should not posses any key and cannot be _Party A_ or _Party B_.

So, basically it means that either _software integrator_ or some other person (e.g. **software distributor**, the one organising software delivery in the local network) must be _Party A_ and _Party B_, but for our model it is somewhat unimportant whether they are one or many. So let's assume that _Party A_ and _Party B_ are both _software integrator_.

This means that _software integrator_ possess both keys and delivers final configuration to _software user_.

- If he knows _platform provisioning data_, he can just put it to config.plist before signing, and no separate file is needed.
- If he does not know _platform provisioning data_, it needs to somehow bypass vaulting and OpenCore.efi signing stages. The latter is a problem I cannot find a good solution so far and the suggestions are welcome here.

**Comment by 1alessandro1** [comment]
> If he does not know platform provisioning data, it needs to somehow bypass vaulting and OpenCore.efi signing stages. The latter is a problem I cannot find a good solution so far and the suggestions are welcome here.

What if the owner of platforminfo.plist signs OpenCore.efi with his own private key after the software integrator has provided its valuted configuration with everything except platforminfo.plist? 

Could it be possble to implement something signed by more than one party?

**Comment by vit9696** [comment]
Well, not quite. Only **OpenCore.efi** can verify _PlatformInfo.plist_ and in the UEFI world there is no safe place to store the keys but inside the binary. For this reason it is not possible to allow somebody adding his key there without tampering with _OpenCore.efi_. And if we require OpenCore.efi modification to be made by the user, the user can resign the whole vault anyway.

Strictly speaking, it may still be useful to store PlatformInfo.plist separately even when the same party signs it, just for the sense of sharing and config.plist verification (equivalence). Perhaps it is worth a try to just ignore the PKI issue and implement this?

**Comment by Qonfused** [comment]
I don't believe the PKI issue is really that applicable as sensitive platform info is machine-specific (unique MLB/ROM/SN/etc) or would have to be generated per user, and currently there is no separation between any party A and the user w/ vaulting.
